### ISSUE TITLE
Tracing fails for Java builds on Mac M1

### ISSUE THREAD

**Issue by wyu0hop** [github/codeql-cli-binaries#101]
hello,i try to use "codeql database create ./test_java --language=java" to build database,but it have errors,like:
```
[2021-12-14 17:20:41] [build-stderr] dyld[4883]: terminating because inserted dylib '/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/libtrace.dylib' could not be loaded: tried: '/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/libtrace.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/local/lib/libtrace.dylib' (no such file), '/usr/lib/libtrace.dylib' (no such file)
[2021-12-14 17:20:41] [build-stderr] dyld[4883]: tried: '/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/libtrace.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/local/lib/libtrace.dylib' (no such file), '/usr/lib/libtrace.dylib' (no such file)
[2021-12-14 17:20:41] [build-stderr] /Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/java/tools/autobuild.sh: line 7:  4883 Abort trap: 6           "${CODEQL_JAVA_HOME}/bin/java" -jar "$CODEQL_EXTRACTOR_JAVA_ROOT/tools/autobuild-fat.jar" autoBuild --no-indexing
[2021-12-14 17:20:41] [ERROR] Spawned process exited abnormally (code 134; tried to run: [/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/preload_tracer, /Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/java/tools/autobuild.sh])
A fatal error occurred: Exit status 134 from command: [/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/java/tools/autobuild.sh]
```
should i rebuild codeql on mac m1?

**Comment by hmakholm** [comment]
Thanks for the report. Is your CodeQL CLI updated to release 2.7.3? There were some improvements in Mac M1 support in that release which _ought_ to have meant you shouldn't see that kind of error.

Also be sure that all of the requirements listed in the _2.7.1_ entry of the [changelog](https://github.com/github/codeql-cli-binaries/blob/main/CHANGELOG.md) (search for "Apple Silicon") are met.

**Comment by binganao** [comment]
Yes, I encountered the same problem, but it was successful in Python

**Comment by on-theway** [comment]
Yes, I have also encountered the same problem.
Version：
` xxxxx  ~ ： codeql --version                                                                                                                                            2 ↵  3501  15:38:51
CodeQL command-line toolchain release 2.7.3.
Copyright (C) 2019-2021 GitHub, Inc.
Unpacked in: /Users/xxx/codeql/codeql
   Analysis results depend critically on separately distributed query and
   extractor modules. To list modules that are visible to the toolchain,
   use 'codeql resolve qlpacks' and 'codeql resolve languages'.`


`：codeql database create /Users/xxx/codeql/CodeQL_java_test --source-root=/Users/xxx/JAVA/java_test --language=java                    2 ↵  3481  15:15:01
Initializing database at /Users/xxx/codeql/CodeQL_java_test.
Running build command: [/Users/xxx/codeql/codeql/java/tools/autobuild.sh]
[2021-12-15 15:38:51] [build-stderr] dyld: could not load inserted library '/Users/xxx/codeql/codeql/tools/osx64/libtrace.dylib' because no suitable image found.  Did find:
[2021-12-15 15:38:51] [build-stderr] 	/Users/xxx/codeql/codeql/tools/osx64/libtrace.dylib: mach-o, but wrong architecture
[2021-12-15 15:38:51] [build-stderr] 	/Users/xxx/codeql/codeql/tools/osx64/libtrace.dylib: stat() failed with errno=1
[2021-12-15 15:38:51] [build-stderr] /Users/xxx/codeql/codeql/java/tools/autobuild.sh: line 7: 14734 Abort trap: 6           "${CODEQL_JAVA_HOME}/bin/java" -jar "$CODEQL_EXTRACTOR_JAVA_ROOT/tools/autobuild-fat.jar" autoBuild --no-indexing
[2021-12-15 15:38:51] [ERROR] Spawned process exited abnormally (code 134; tried to run: [/Users/xxx/codeql/codeql/tools/osx64/preload_tracer, /Users/xxx/codeql/codeql/java/tools/autobuild.sh])
A fatal error occurred: Exit status 134 from command: [/Users/xxx/codeql/codeql/java/tools/autobuild.sh]`

**Comment by binganao** [comment]
Well, I've successfully solved the problem that happened to me. I need to execute while creating the database
```--command="mvn clean install --file pom.xml"```

**Comment by hmakholm** [comment]
Thanks -- it looks like we have a problem a problem with building Java databases on M1 macs. We have some experts looking at it now.

**Comment by hmakholm** [comment]
This problem should be fixed in the soon-to-be-released version 2.7.4 of the CLI.

**Comment by hmakholm** [comment]
I spoke too soon! The fix we have works, but the release that would contain it has been postponed to January due to unrelated test failures (and holidays). :-(

Meanwhile, chances are fairly good that rolling back to release 2.7.1 would unblock you -- the underlying problem exists only in 2.7.2 and 2.7.3.

**Comment by binganao** [comment]
> 我说话太快了！我们有修复程序，但由于测试失败（和节假日）无关，包含它的版本已推迟到1月。:-(
> 
> 与此同时，回滚到2.7.1版本可能会解除对您的封锁——根本问题仅在2.7.2和2.7.3中存在。

Thank you so much

**Comment by binganao** [comment]
> I spoke too soon! The fix we have works, but the release that would contain it has been postponed to January due to unrelated test failures (and holidays). :-(
> 
> Meanwhile, chances are fairly good that rolling back to release 2.7.1 would unblock you -- the underlying problem exists only in 2.7.2 and 2.7.3.

Thank you so much

**Comment by Oldman19** [comment]
> 您好，我尝试使用“codeql database create ./test_java --language=java”来构建数据库，但它有错误，例如：
> 
> ```
> [2021-12-14 17:20:41] [build-stderr] dyld[4883]: terminating because inserted dylib '/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/libtrace.dylib' could not be loaded: tried: '/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/libtrace.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/local/lib/libtrace.dylib' (no such file), '/usr/lib/libtrace.dylib' (no such file)
> [2021-12-14 17:20:41] [build-stderr] dyld[4883]: tried: '/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/libtrace.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/local/lib/libtrace.dylib' (no such file), '/usr/lib/libtrace.dylib' (no such file)
> [2021-12-14 17:20:41] [build-stderr] /Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/java/tools/autobuild.sh: line 7:  4883 Abort trap: 6           "${CODEQL_JAVA_HOME}/bin/java" -jar "$CODEQL_EXTRACTOR_JAVA_ROOT/tools/autobuild-fat.jar" autoBuild --no-indexing
> [2021-12-14 17:20:41] [ERROR] Spawned process exited abnormally (code 134; tried to run: [/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/preload_tracer, /Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/java/tools/autobuild.sh])
> A fatal error occurred: Exit status 134 from command: [/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/java/tools/autobuild.sh]
> ```
> 
> 我应该在 mac m1 上重建 codeql 吗？

Brother, I also encountered the same problem, how did you solve it.

**Comment by Oldman19** [comment]
> hello,i try to use "codeql database create ./test_java --language=java" to build database,but it have errors,like:
> 
> ```
> [2021-12-14 17:20:41] [build-stderr] dyld[4883]: terminating because inserted dylib '/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/libtrace.dylib' could not be loaded: tried: '/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/libtrace.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/local/lib/libtrace.dylib' (no such file), '/usr/lib/libtrace.dylib' (no such file)
> [2021-12-14 17:20:41] [build-stderr] dyld[4883]: tried: '/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/libtrace.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/local/lib/libtrace.dylib' (no such file), '/usr/lib/libtrace.dylib' (no such file)
> [2021-12-14 17:20:41] [build-stderr] /Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/java/tools/autobuild.sh: line 7:  4883 Abort trap: 6           "${CODEQL_JAVA_HOME}/bin/java" -jar "$CODEQL_EXTRACTOR_JAVA_ROOT/tools/autobuild-fat.jar" autoBuild --no-indexing
> [2021-12-14 17:20:41] [ERROR] Spawned process exited abnormally (code 134; tried to run: [/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/tools/osx64/preload_tracer, /Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/java/tools/autobuild.sh])
> A fatal error occurred: Exit status 134 from command: [/Users/xxxxxx/Desktop/learning/codeql/codeql-home/codeql/java/tools/autobuild.sh]
> ```
> 
> should i rebuild codeql on mac m1?

Brother, I also encountered the same problem, how did you solve it.

**Comment by safe6Sec** [comment]
@hmakholm I also faced the same problem with the latest version 2.12.2

```
[2023-02-14 21:37:26] [build-stdout] [2023-02-14 21:37:26] [autobuild] dyld[28217]: terminating because inserted dylib '/Users/safe6sec/code/codeql/codeql/tools/osx64/libtrace.dylib' could not be loaded: tried: '/Users/safe6sec/code/codeql/codeql/tools/osx64/libtrace.dylib' (fat file, but missing compatible architecture (have 'x86_64,arm64', need 'arm64e')), '/usr/local/lib/libtrace.dylib' (no such file), '/usr/lib/libtrace.dylib' (no such file)
[2023-02-14 21:37:26] [build-stdout] [2023-02-14 21:37:26] [autobuild] dyld[28217]: tried: '/Users/safe6sec/code/codeql/codeql/tools/osx64/libtrace.dylib' (fat file, but missing compatible architecture (have 'x86_64,arm64', need 'arm64e')), '/usr/local/lib/libtrace.dylib' (no such file), '/usr/lib/libtrace.dylib' (no such file)
[2023-02-14 21:37:26] [ERROR] Spawned process exited abnormally (code 134; tried to run: [/Users/safe6sec/code/codeql/codeql/tools/osx64/preload_tracer, /Users/safe6sec/code/codeql/codeql/java/tools/autobuild.sh])
```

**Comment by hmakholm** [comment]
@safe6Sec This is a known problem. Unfortunately we can't build working `arm64e` tracer binaries due to Apple's signing policies. Some users have found that a workaround that works for them is to run the database creation step as `arch -x86_64 codeql database create ...`

**Comment by safe6Sec** [comment]
> @safe6Sec This is a known problem. Our experience is that it only seems to happen when System Integrity Protection in MacOS is _disabled_, but I don't know the exact reason why that triggers it.
> 
> Unfortunately we can't build working `arm64e` tracer binaries due to Apple's signing policies. Some users have found that a workaround that works for them is to run the database creation step as `arch -x86_64 codeql database create ...`

Thank you

**Comment by AlexDenisov** [comment]
Hi folks, I believe we've fixed this issue in the latest release (2.16.3). We'd appreciate if you can give it a try and share feedback on whether it works for you or not 🙌
